lib/jsonapi/plugs/format_required.ex

```diff
@@ -12,6 +12,57 @@ defmodule JSONAPI.FormatRequired do
 
   def call(%{method: method} = conn, _opts) when method in ~w[DELETE GET HEAD], do: conn
 
+  def call(
+        %{method: method, params: %{"data" => %{"type" => _, "relationships" => relationships}}} =
+          conn,
+        _
+      )
+      when method in ~w[POST PATCH] and not is_map(relationships) do
+    send_error(conn, relationships_missing_object())
+  end
+
+  def call(
+        %{
+          method: method,
+          params: %{"data" => %{"type" => _, "relationships" => relationships}}
+        } = conn,
+        _
+      )
+      when method in ~w[POST PATCH] and is_map(relationships) do
+    errors =
+      Enum.reduce(relationships, [], fn
+        {_relationship_name, %{"data" => %{"type" => _type, "id" => _}}}, acc ->
+          acc
+
+        {relationship_name, %{"data" => %{"type" => _type}}}, acc ->
+          error = missing_relationship_data_id_param_error(relationship_name)
+          [error | acc]
+
+        {relationship_name, %{"data" => %{"id" => _type}}}, acc ->
+          error = missing_relationship_data_type_param_error(relationship_name)
+          [error | acc]
+
+        {relationship_name, %{"data" => %{}}}, acc ->
+          id_error = missing_relationship_data_id_param_error(relationship_name)
+          type_error = missing_relationship_data_type_param_error(relationship_name)
+          [id_error | [type_error | acc]]
+
+        {_relationship_name, %{"data" => _}}, acc ->
+          # Allow things other than resource identifier objects per https://jsonapi.org/format/#document-resource-object-linkage
```

Comment:
Mostly I don't have a strong personal need to try and validate the list / null cases super thoroughly right now so I left them as a passthrough for now but presumably those could be validated eventually as well.

Reply:
I don't think you need to add all related validations for this PR to be a win, but could you mention the unvalidated cases by name in this comment? I know what it means, but it kind of reads as "the spec allows all other structures here" instead of "there are a couple of other valid structures here".

Reply:
Copy/pasted comments with the 3 (of 4) other allowed object shapes from the docs